src/lavinmq/mqtt/broker.cr

```diff
@@ -0,0 +1,111 @@
+require "./retain_store"
+require "../vhost"
+
+module LavinMQ
+  module MQTT
+    class Broker
+      getter vhost, sessions
+
+      # The `Broker` class acts as an intermediary between the `Server` and MQTT connections.
+      # It is initialized by the `Server` and manages client connections, sessions, and message exchange.
+      # Responsibilities include:
+      # - Handling client connections and disconnections
+      # - Managing client sessions (clean and persistent)
+      # - Publishing messages to the exchange
+      # - Subscribing and unsubscribing clients to/from topics
+      # - Handling the retain store
+      # - Interfacing with the virtual host (vhost) and the exchange to route messages
+      # The `Broker` class helps keep the MQTT client concise and focused on the protocol.
+
+      def initialize(@vhost : VHost, @replicator : Clustering::Replicator)
+        @sessions = Sessions.new(@vhost)
+        @clients = Hash(String, Client).new
+        @retain_store = RetainStore.new(File.join(@vhost.data_dir, "mqtt_reatined_store"), @replicator)
+        @exchange = MQTT::Exchange.new(@vhost, EXCHANGE, @retain_store)
+        @vhost.exchanges[EXCHANGE] = @exchange
+      end
+
+      def session_present?(client_id : String, clean_session) : Bool
+        return false if clean_session
+        session = sessions[client_id]?
+        return false if session.nil? || session.clean_session?
+        true
+      end
+
+      def add_client(socket, connection_info, user, packet)
+        if prev_client = @clients[packet.client_id]?
+          prev_client.close("New client #{connection_info.src} (username=#{packet.username}) connected as #{packet.client_id}")
+        end
+        client = MQTT::Client.new(socket,
+                                  connection_info,
+                                  user,
+                                  self,
+                                  packet.client_id,
+                                  packet.clean_session?,
+                                  packet.keepalive,
+                                  packet.will)
+        if session = sessions[client.client_id]?
+          if client.clean_session?
+            sessions.delete session
+          else
+            session.client = client
+          end
+        end
+        @clients[packet.client_id] = client
+        client
+      end
+
+      def remove_client(client)
+        client_id = client.client_id
+        if session = sessions[client_id]?
+          session.client = nil
+          sessions.delete(client_id) if session.clean_session?
+        end
+        @clients.delete client_id
+        vhost.rm_connection(client)
+      end
+
+      def publish(packet : MQTT::Publish)
+        @exchange.publish(packet)
+      end
+
+      def subscribe(client, packet)
+        unless session = sessions[client.client_id]?
+          session = sessions.declare(client.client_id, client.@clean_session)
+          session.client = client
+        end
```

Comment:
a bit counterintuitive to have an "unless" assignment.
 
```suggestion
        session = sessions[client.client_id]? || begin
          s = sessions.declare(client.client_id, client.clean_session)
          s.client = client
          s
        end
```

also this seems to be the only place where `Session#declare` is called, so why not pass in the whole client object? in which case it could look like this:

```suggestion
        session = sessions[client.client_id]? || sessions.declare(client)
```